### PR TITLE
feat: optimized chat list rendering without full re-renders #1156

### PR DESCRIPTION
**Pull request by alexis-kochetov**

### CODE REVIEW

File: src/dotnet/Chat.UI.Blazor/Components/ChatList/ChatList.razor [hunk 1]
```diff
@@ -1,28 +1,67 @@
...
-    }
-</div>
-@if (Kind == ChatListKind.Pinned) {
-    <div class="mx-4 mb-2 border-b border-bg-04"></div>
+@if (isLoaded) {
+    <div class="chat-list chat-list-all">
+        @for (var i = 0; i < chatCount; i++) {
+            var position = i;
+
+            <ChatListItem @key="@position" Position="@position" ListKind="@ChatListKind.All" />
```
Comment (frolyo) [inline src/dotnet/Chat.UI.Blazor/Components/ChatList/ChatList.razor:16]: why is index a key?

File: src/dotnet/Chat.UI.Blazor/Components/ChatList/ChatListItem.razor [hunk 2]
```diff
@@ -1,101 +1,147 @@
...
-                                <AuthorName Class="text-xs text-02 font-medium" AuthorSid="@lastTextEntry.AuthorId" ShowDetailsOnClick="false"/>:
-                            </span>
-                        }
-                        @info.LastTextEntryText
+@if (m == Model.Loading) {
+    <ChatListItemSkeleton />
+}
+else {
+    <NavbarItem
+        Url="@($"/chat/{chat.Id}")"
```
Comment (frolyo) [inline src/dotnet/Chat.UI.Blazor/Components/ChatList/ChatListItem.razor:21]: why not `Links.Chat(chat.Id)`?

File: src/dotnet/Chat.UI.Blazor/Components/ChatList/ChatListItem.razor [hunk 3]
```diff
@@ -1,101 +1,147 @@
...
+        IsSelected="@chatState.IsSelected"
+        data-menu="@(MenuRef.New<ChatMenu>(chat.Id, ListKind.Format()).ToString())"
+        data-menu-placement="@(FloatingPosition.BottomStart.ToPositionString())">
+        <ChildContent>
+            <div class="flex-1 flex-x items-center gap-x-2" @onclick="()=>OnClick(chat.Id)">
```
Comment (frolyo) [inline src/dotnet/Chat.UI.Blazor/Components/ChatList/ChatListItem.razor:26]: lost spaces in onclick

File: src/dotnet/Chat.UI.Blazor/Components/ChatList/ChatListItem.razor [hunk 4]
```diff
@@ -1,101 +1,147 @@
...
-    {
-        var chatId = await ChatUI.FixChatId(ChatInfo.Id, cancellationToken).ConfigureAwait(false);
-        var chatState = await ChatUI.GetState(chatId, false, cancellationToken).ConfigureAwait(false);
-        return chatState ?? ChatState.None;
+    protected override async Task<Model> ComputeState(CancellationToken cancellationToken) {
+        ChatInfo? chatInfo;
+        if (ListKind == ChatListKind.Active)
+            chatInfo = ChatInfo;
+        else {
+            var chatId = await ChatUI.GetChatId(Position, cancellationToken).ConfigureAwait(false);
```
Comment (frolyo) [inline src/dotnet/Chat.UI.Blazor/Components/ChatList/ChatListItem.razor:117]: afair we agreed not to use ConfigureAwait(false) in components

File: src/dotnet/Chat.UI.Blazor/Components/ChatList/ChatListSortButton.razor [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+            SortOrder = settings.Order,
+        };
+    }
+
+    public sealed record Model {
+        public static Model Loading { get; } = new();
+
+        public ChatListOrder SortOrder { get; init; }
+    }
+
```
Comment (frolyo) [inline src/dotnet/Chat.UI.Blazor/Components/ChatList/ChatListSortButton.razor:38]: excess line

File: src/dotnet/Chat.UI.Blazor/Components/ChatList/ChatListTabUnreadCount.razor [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        History.NavigateTo(Links.Chat(unreadChatId.Value));
+    }
+
+    public sealed record Model {
+        public static Model Loading { get; } = new();
+
+        public Trimmed<int> Count { get; init; }
+        public ChatInfo? FirstUnread { get; init; }
+    }
+
```
Comment (frolyo) [inline src/dotnet/Chat.UI.Blazor/Components/ChatList/ChatListTabUnreadCount.razor:52]: redundant line

**Review by frolyo — APPROVED** [review]

File: src/dotnet/Chat.UI.Blazor/Components/ChatList/ActiveChatList.razor [hunk 7]
```diff
@@ -31,10 +41,10 @@
     protected override async Task<Model> ComputeState(CancellationToken cancellationToken) {
         var activeChats = await ActiveChatsUI.ActiveChats.Use(cancellationToken);
         var chats = (await activeChats
-            .OrderByDescending(c => c.Recency)
-            .Select(c => ChatUI.Get(c.ChatId, cancellationToken))
-            .Collect()
-            ).SkipNullItems()
+                .OrderByDescending(c => c.Recency)
```
Comment (alexyakunin) [inline src/dotnet/Chat.UI.Blazor/Components/ChatList/ActiveChatList.razor:44]: IMO it's fine to column-format such stuff, i.e. w/o an extra indent.

File: src/dotnet/Chat.UI.Blazor/Components/ChatList/ChatList.razor (hunk 1 above)
Comment (alexyakunin) [reply src/dotnet/Chat.UI.Blazor/Components/ChatList/ChatList.razor:16]: Coz the idea is to make sure that only moved / updated items are re-rendered.

Comment (alexyakunin) [reply src/dotnet/Chat.UI.Blazor/Components/ChatList/ChatList.razor:16]: Technically key isn't needed here, btw. And maybe it's even better to do it w/o a key, coz having key = chatCount boxings per render.

Comment (alexyakunin) [reply src/dotnet/Chat.UI.Blazor/Components/ChatList/ChatList.razor:16]: I.e. w/o a key Blazor will simply map these components by index.

File: src/dotnet/Chat.UI.Blazor/Components/ChatList/ChatList.razor [hunk 8]
```diff
@@ -1,28 +1,67 @@
...
+            InitialValue = Model.Loading,
+            UpdateDelayer = FixedDelayer.Instant,
+        };
+
+    protected override async Task<Model> ComputeState(CancellationToken cancellationToken) {
+        var tracer = ReferenceEquals(State.Value, Model.Loading)
+            ? Tracer[$"{nameof(ChatListNavbarWidget)}.{nameof(ComputeState)}"]
+            : Tracer.None;
+
+        var region = tracer.Region("ChatUI.GetChatCount");
```
Comment (alexyakunin) [inline src/dotnet/Chat.UI.Blazor/Components/ChatList/ChatList.razor:44]: Why you guys just don't use `using var _ =`? `new(...)` in the end doesn't change much

File: src/dotnet/Chat.UI.Blazor/Components/ChatList/ChatListItem.razor (hunk 2 above)
Comment (alexyakunin) [reply src/dotnet/Chat.UI.Blazor/Components/ChatList/ChatListItem.razor:21]: Yep.

File: src/dotnet/Chat.UI.Blazor/Components/ChatList/ChatListItem.razor (hunk 4 above)
Comment (alexyakunin) [reply src/dotnet/Chat.UI.Blazor/Components/ChatList/ChatListItem.razor:117]: Yep.

File: src/dotnet/Chat.UI.Blazor/Components/ChatList/ChatListItem.razor [hunk 9]
```diff
@@ -1,101 +1,147 @@
...
+            if (!chatId.HasValue)
+                return Model.None;
+            chatInfo = await ChatUI.Get(chatId.Value, cancellationToken).ConfigureAwait(false);
```
Comment (alexyakunin) [inline src/dotnet/Chat.UI.Blazor/Components/ChatList/ChatListItem.razor:120]: Same here & further.

File: src/dotnet/Chat.UI.Blazor/Components/ChatList/ChatListTabUnreadCount.razor [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+    if (m == Model.Loading || count <= 0)
+        return;
+}
+
+<div class="c-badge">
+    <UnreadCount Value="@count" Click="() => OnUnreadBadgeClick(firstUnread?.Id)"/>
+</div>
+
+@code {
+
```
Comment (alexyakunin) [inline src/dotnet/Chat.UI.Blazor/Components/ChatList/ChatListTabUnreadCount.razor:17]: Extra empty line.

File: src/dotnet/Chat.UI.Blazor/Components/ChatList/ChatListTabUnreadCount.razor [hunk 11]
```diff
@@ -0,0 +1,53 @@
+@namespace ActualChat.Chat.UI.Blazor.Components
+@inherits ComputedStateComponent<ChatListTabUnreadCount.Model>
+
```
Comment (alexyakunin) [inline src/dotnet/Chat.UI.Blazor/Components/ChatList/ChatListTabUnreadCount.razor:3]: IMO doesn't make sense to leave an extra empty line here too.

File: src/dotnet/Chat.UI.Blazor/Services/ChatUI.StateSync.cs [hunk 12]
```diff
@@ -40,6 +41,56 @@ private async Task InvalidateSelectedChatDependencies(CancellationToken cancella
         }
     }
 
+    private async Task MaintainChatPositions(CancellationToken cancellationToken)
+    {
+        var settingsChanges = ListSettings.Changes(cancellationToken).ConfigureAwait(false);
+        await foreach (var cSettingsChange in settingsChanges.ConfigureAwait(false)) {
+            var settings = cSettingsChange.Value;
+            var filterId = settings.Filter.Id;
+            var cListFiltered = await Computed
```
Comment (alexyakunin) [inline src/dotnet/Chat.UI.Blazor/Services/ChatUI.StateSync.cs:50]: This thing can be captured outside of `foreach` loop.

Comment (alexyakunin) [reply src/dotnet/Chat.UI.Blazor/Services/ChatUI.StateSync.cs:50]: You might just call `computed = await computed.Update()` inside.

File: src/dotnet/Chat.UI.Blazor/Services/ChatUI.StateSync.cs [hunk 13]
```diff
@@ -40,6 +41,56 @@ private async Task InvalidateSelectedChatDependencies(CancellationToken cancella
         }
     }
 
+    private async Task MaintainChatPositions(CancellationToken cancellationToken)
```
Comment (alexyakunin) [inline src/dotnet/Chat.UI.Blazor/Services/ChatUI.StateSync.cs:44]: The name is misleading. `InvalidateChatPositions`?

File: src/dotnet/Chat.UI.Blazor/Services/ChatUI.StateSync.cs [hunk 14]
```diff
@@ -40,6 +41,56 @@ private async Task InvalidateSelectedChatDependencies(CancellationToken cancella
...
+
+                    _chatPositionMap = result;
+                }
+                if (positionsToInvalidate.Count <= 0)
+                    continue;
+
+                Log.LogDebug("MaintainChatPositions: invalidate {Count} items", positionsToInvalidate.Count);
+                using (Computed.Invalidate())
+                    foreach (var i in positionsToInvalidate)
+                        _ = GetChatId(i, default);
```
Comment (alexyakunin) [inline src/dotnet/Chat.UI.Blazor/Services/ChatUI.StateSync.cs:88]: I'd rather rename this method to `GetChatIdByPosition`, btw.

File: src/dotnet/Chat.UI.Blazor/Services/ChatUI.StateSync.cs (hunk 13 above)
Comment (alexyakunin) [reply src/dotnet/Chat.UI.Blazor/Services/ChatUI.StateSync.cs:44]: Maybe `InvalidateGetChatIdByPosition`?

Comment (alexyakunin) [reply src/dotnet/Chat.UI.Blazor/Services/ChatUI.StateSync.cs:44]: IMO you shouldn't break naming conventions other ppl use - see how other methods are named in such classes, `InvalidateXxx` is a very common pattern.

File: src/dotnet/Chat.UI.Blazor/Services/ChatUI.cs [hunk 15]
```diff
@@ -96,40 +101,46 @@ void INotifyInitialized.Initialized()
...
     [ComputeMethod]
-    protected virtual async Task<IReadOnlyList<ChatInfo>> ListFiltered(Symbol filterId, CancellationToken cancellationToken)
+    public virtual Task<ChatId?> GetChatId(int position, CancellationToken cancellationToken = default)
+    {
+        ChatId? chatId;
+        lock (_chatPositionLock) {
+            if (position >= _chatPositionMap.Count)
+                return Task.FromResult<ChatId?>(null);
+
+            chatId = _chatPositionMap[position];
```
Comment (alexyakunin) [inline src/dotnet/Chat.UI.Blazor/Services/ChatUI.cs:119]: `_chatPositionMap` is an amazing name for the list.

Maybe `_chatIdByPosition`?

Comment (alexyakunin) [reply src/dotnet/Chat.UI.Blazor/Services/ChatUI.cs:119]: IMO event `_chatIds` would be a better choice...

File: src/dotnet/Chat.UI.Blazor/Services/ChatUI.cs [hunk 16]
```diff
@@ -96,40 +101,46 @@ void INotifyInitialized.Initialized()
...
+        var filteredChats = await ListFiltered(settings.Filter.Id, cancellationToken).ConfigureAwait(false);
+        return filteredChats.Count;
     }
 
     [ComputeMethod]
-    protected virtual async Task<IReadOnlyList<ChatInfo>> ListFiltered(Symbol filterId, CancellationToken cancellationToken)
+    public virtual Task<ChatId?> GetChatId(int position, CancellationToken cancellationToken = default)
+    {
+        ChatId? chatId;
+        lock (_chatPositionLock) {
```
Comment (alexyakunin) [inline src/dotnet/Chat.UI.Blazor/Services/ChatUI.cs:115]: One other thing is: IMO it's fine to lock on private object, and here you can use the list itself - not sure why it's not a read-only list, btw, I'd rather convert it to a read-only one.

File: src/dotnet/Chat.UI.Blazor/Services/ChatUI.cs [hunk 17]
```diff
@@ -96,40 +101,46 @@ void INotifyInitialized.Initialized()
...
-        var result = orderedChats.ToList();
-        DebugLog?.LogDebug("<- List: {Settings}", settings);
-        return result;
+        var filteredChats = await ListFiltered(settings.Filter.Id, cancellationToken).ConfigureAwait(false);
+        return filteredChats.Count;
     }
 
     [ComputeMethod]
-    protected virtual async Task<IReadOnlyList<ChatInfo>> ListFiltered(Symbol filterId, CancellationToken cancellationToken)
+    public virtual Task<ChatId?> GetChatId(int position, CancellationToken cancellationToken = default)
```
Comment (alexyakunin) [inline src/dotnet/Chat.UI.Blazor/Services/ChatUI.cs:112]: You can return simply `ChatId` assuming that it anyway can be `IsNone`.

File: src/dotnet/Chat.UI.Blazor/Services/ChatUI.cs [hunk 18]
```diff
@@ -96,40 +101,46 @@ void INotifyInitialized.Initialized()
...
+    {
+        ChatId? chatId;
+        lock (_chatPositionLock) {
+            if (position >= _chatPositionMap.Count)
+                return Task.FromResult<ChatId?>(null);
```
Comment (alexyakunin) [inline src/dotnet/Chat.UI.Blazor/Services/ChatUI.cs:117]: Am I right that `null` items simply aren't rendered?

File: src/dotnet/Chat.UI.Blazor/Services/ChatUI.cs [hunk 19]
```diff
@@ -351,6 +367,30 @@ public async ValueTask<SyncedStateLease<ChatPosition>> LeaseReadPositionState(Ch
 
     // Private methods
 
+    private IReadOnlyList<ChatInfo> SortChats(IReadOnlyCollection<(ChatInfo ChatInfo, double Rank)> rankedChats, ChatListOrder order)
```
Comment (alexyakunin) [inline src/dotnet/Chat.UI.Blazor/Services/ChatUI.cs:370]: `IReadOnlyCollection` -> `IEnumerable`? `Count` isn't used anyway.

File: src/dotnet/Core/ComputedExt.cs [hunk 20]
```diff
@@ -41,4 +41,31 @@ public static class ComputedExt
             cts.CancelAndDisposeSilently();
         }
     }
+
+    public static async IAsyncEnumerable<Computed<T>> Until<T>(
```
Comment (alexyakunin) [inline src/dotnet/Core/ComputedExt.cs:45]: Updated & moved this one to `AsyncEnumerableExt` already.

**Review by alexyakunin — COMMENTED** [review]
Hi, overall, IMO this is exactly what we need. But there are tons of minor comments ;)

File: src/dotnet/Chat.UI.Blazor/Components/ChatList/ChatList.razor (hunk 1 above)
Comment (alexis-kochetov) [reply src/dotnet/Chat.UI.Blazor/Components/ChatList/ChatList.razor:16]: I will try to get rid of it

File: src/dotnet/Chat.UI.Blazor/Components/ChatList/ChatListItem.razor (hunk 2 above)
Comment (alexis-kochetov) [reply src/dotnet/Chat.UI.Blazor/Components/ChatList/ChatListItem.razor:21]: I have copied the code without reading)))

Comment (alexis-kochetov) [reply src/dotnet/Chat.UI.Blazor/Components/ChatList/ChatListItem.razor:21]: Just moved it to another location

File: src/dotnet/Chat.UI.Blazor/Components/ChatList/ActiveChatList.razor (hunk 7 above)
Comment (alexis-kochetov) [reply src/dotnet/Chat.UI.Blazor/Components/ChatList/ActiveChatList.razor:44]: I like indents!!!!!1111)))

File: src/dotnet/Chat.UI.Blazor/Services/ChatUI.StateSync.cs (hunk 13 above)
Comment (alexis-kochetov) [reply src/dotnet/Chat.UI.Blazor/Services/ChatUI.StateSync.cs:44]: It's not just invalidating! It updates the list as well

File: src/dotnet/Chat.UI.Blazor/Services/ChatUI.cs (hunk 15 above)
Comment (alexis-kochetov) [reply src/dotnet/Chat.UI.Blazor/Services/ChatUI.cs:119]: Don't think that each map is a dictionary

File: src/dotnet/Chat.UI.Blazor/Services/ChatUI.cs (hunk 16 above)
Comment (alexis-kochetov) [reply src/dotnet/Chat.UI.Blazor/Services/ChatUI.cs:115]: Because we are replacing it

**Pull request merged**